version/manager/utils.go

```diff
@@ -0,0 +1,103 @@
+/*
+ *    Copyright 2018 Insolar
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package manager
+
+import (
+	"github.com/blang/semver"
+	"github.com/insolar/insolar/core"
+	"github.com/insolar/insolar/log"
+	"github.com/insolar/insolar/version"
+	"github.com/pkg/errors"
+)
+
+func ProcessVersionConsensus(nodes []*core.ActiveNode) error {
+	if len(nodes) == 0 {
+		return errors.New("List of nodes is empty")
+	}
+	mapOfVersions := getMapOfVersion(nodes)
+	topVersion, err := getMaxVersion(getRequired(len(nodes)), mapOfVersions)
+	if err != nil {
+		return err
+	}
+	if topVersion != nil {
+		vm, err := GetVersionManager()
+		if err != nil {
+			return err
+		}
+		currentVersion, err := ParseVersion(version.Version)
+		if err != nil {
+			return errors.New("current version is invalid: " + err.Error())
+		}
+		if currentVersion.Compare(*topVersion) != 0 {
+			log.Warn("WARNING! Current version: " + StringVersion(currentVersion) + ", must go to version: " + StringVersion(topVersion))
+		}
+		vm.AgreedVersion = topVersion
+	}
+	return nil
+}
+
+func getMapOfVersion(nodes []*core.ActiveNode) *map[string]int {
+	mapOfVersions := make(map[string]int)
+
+	// ToDo: I Need a version from the ActiveNodeList
+	for _ = range nodes { // node
+		ver := "v0.5.0" // node.Version
+		if _, ok := mapOfVersions[ver]; ok {
+			mapOfVersions[ver]++
+		} else {
+			mapOfVersions[ver] = 1
+		}
+	}
+	return &mapOfVersions
+}
+
+func getMaxVersion(required int, mapOfVersions *map[string]int) (*semver.Version, error) {
```

Comment:
I think there's a problem here.
If for example we have four nodes with v0.5.0, v0.5.1, v0.5.2, v0.5.3, function will return nil but the correct answer is v0.5.0.

Reply:
Changed